src/core/qgsnetworkaccessmanager.h

```diff
@@ -275,30 +370,66 @@ class CORE_EXPORT QgsNetworkAccessManager : public QNetworkAccessManager
      */
     void downloadProgress( int requestId, qint64 bytesReceived, qint64 bytesTotal );
 
+#ifndef QT_NO_SSL
+
+    /**
+     * Emitted when a network request encounters SSL \a errors.
+     *
+     * The \a requestId argument reflects the unique ID identifying the original request which the progress report relates to.
+     *
+     * This signal is propagated to the main thread QgsNetworkAccessManager instance, so it is necessary
+     * only to connect to the main thread's signal in order to receive notifications about SSL errors
+     * from any thread.
+     *
+     * \since QGIS 3.6
+     */
+    void requestEncounteredSslErrors( int requestId, const QList<QSslError> &errors );
+
+#ifndef SIP_RUN
+///@cond PRIVATE
+    // these signals are for internal use only - it's not safe to connect by external code
+    void sslErrorsOccurred( QNetworkReply *, const QList<QSslError> &errors );
+    void sslErrorsHandled( QNetworkReply *reply );
+///@endconf
+#endif
+
+#endif
+
     /**
      * \deprecated Use the thread-safe requestAboutToBeCreated( QgsNetworkRequestParameters ) signal instead.
      */
     Q_DECL_DEPRECATED void requestCreated( QNetworkReply * ) SIP_DEPRECATED;
 
     void requestTimedOut( QNetworkReply * );
 
+
   private slots:
     void abortRequest();
 
     void onReplyFinished( QNetworkReply *reply );
 
     void onReplyDownloadProgress( qint64 bytesReceived, qint64 bytesTotal );
+#ifndef QT_NO_SSL
+    void onReplySslErrors( const QList<QSslError> &errors );
+#endif
+
+    void afterSslErrorHandled( QNetworkReply *reply );
 
   protected:
     QNetworkReply *createRequest( QNetworkAccessManager::Operation op, const QNetworkRequest &req, QIODevice *outgoingData = nullptr ) override;
 
   private:
+    void unlockAfterSslErrorHandled();
+
     QList<QNetworkProxyFactory *> mProxyFactories;
     QNetworkProxy mFallbackProxy;
     QStringList mExcludedURLs;
     bool mUseSystemProxy = false;
     bool mInitialized = false;
     static QgsNetworkAccessManager *sMainNAM;
+    std::unique_ptr< QgsSslErrorHandler > mSslErrorHandler;
+    QMutex mMutex;
+    QWaitCondition mSslWaitCondition;
```

Comment:
maybe good to note these two are used only by worker threads?

and maybe s/mMutex/mSslErrorHandlerMutex/ so it clear how it is used?